### PR TITLE
Add/update BBEs for `isolated`

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
This PR
- introduces a separate section for isolation concepts

- adds two BBEs
  - isolated objects - fixes https://github.com/ballerina-platform/ballerina-distribution/issues/913
  - isolated variables - fixes https://github.com/ballerina-platform/ballerina-distribution/issues/935

- updates the isolated functions BBE to include `isolated` variables and improves comments

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: examples/isolated-functions/isolated_functions.bal [hunk 1]
```diff
@@ -27,8 +27,15 @@ isolated class UniqueGreetingsStore {
     }
 }
 
+// A `final` variable of type `UniqueGreetingsStore` which
```
Comment (praneesha) [inline examples/isolated-functions/isolated_functions.bal:30]:
```suggestion
// A `final` variable of the type `UniqueGreetingsStore`, which
```

File: examples/isolated-functions/isolated_functions.bal [hunk 2]
```diff
@@ -27,8 +27,15 @@ isolated class UniqueGreetingsStore {
...
+// is a subtype of `isolated object {}`.
+// This variable can be accessed in an `isolated` function.
 final UniqueGreetingsStore uniqueGreetings = new;
 
+// An `isolated` variable that can be accessed within an
```
Comment (praneesha) [inline examples/isolated-functions/isolated_functions.bal:35]:
```suggestion
// An `isolated` variable can be accessed within an
```

File: examples/isolated-functions/isolated_functions.bal [hunk 3]
```diff
@@ -27,8 +27,15 @@ isolated class UniqueGreetingsStore {
...
+// `isolated` function if it does so within a `lock` statement.
```
Comment (praneesha) [inline examples/isolated-functions/isolated_functions.bal:36]:
```suggestion
// `isolated` function if it resides within a `lock` statement.
```

File: examples/isolated-functions/isolated_functions.bal [hunk 4]
```diff
@@ -43,11 +50,16 @@ isolated function addGreetings(string[] greetings, int count) {
     // the mutable array passed as an argument.
     greetings.push(...builtGreetings);
 
-    // Access the `final isolated object {}` variable `uniqueGreetings`,
+    // Access the `final` `isolated object {}` variable `uniqueGreetings`,
```
Comment (praneesha) [inline examples/isolated-functions/isolated_functions.bal:53]:
```suggestion
    // Access the `final` `isolated object {}` variable `uniqueGreetings`
```

File: examples/isolated-functions/isolated_functions.bal [hunk 5]
```diff
@@ -43,11 +50,16 @@ isolated function addGreetings(string[] greetings, int count) {
...
     // and call a method to store the greetings if they are unique.
     foreach string greeting in builtGreetings {
         uniqueGreetings.add(greeting);
     }
+
+    // Access the `isolated` variable within a `lock` statement,
```
Comment (praneesha) [inline examples/isolated-functions/isolated_functions.bal:59]:
```suggestion
    // Access the `isolated` variable within a `lock` statement.
```

File: examples/isolated-functions/isolated_functions.bal [hunk 6]
```diff
@@ -43,11 +50,16 @@ isolated function addGreetings(string[] greetings, int count) {
...
+    lock {
+        updateCount += 1;
+    }
 }
 
 // An `isolated` function, which accesses global immutable state and
```
Comment (praneesha) [inline examples/isolated-functions/isolated_functions.bal:65]:
```suggestion
// An `isolated` function, which accesses the global, immutable state and
```

File: examples/isolated-functions/isolated_functions.bal [hunk 7]
```diff
@@ -75,4 +87,9 @@ public function main() {
     addGreetings(mutableGreetingArray, 2);
 
     io:println(mutableGreetingArray);
+
+    // Access the `isolated` variable within a `lock` statement,
```
Comment (praneesha) [inline examples/isolated-functions/isolated_functions.bal:91]:
```suggestion
    // Access the `isolated` variable within a `lock` statement.
```

File: examples/isolated-functions/isolated_functions.description [hunk 8]
```diff
@@ -1,4 +1,6 @@
-// An `isolated` function is a function that accesses mutable state only via its parameters.
-// Such a function can only access module level state if it is final and the type is a subtype of
-// `readonly` or `isolated object {}`. Any function or method call in an `isolated` function should
-// also be calling an `isolated` function or method.
+// An `isolated` function is a function within which specific rules apply when accessing mutable state.
```
Comment (praneesha) [inline examples/isolated-functions/isolated_functions.description:1]:
```suggestion
// An `isolated` function is a function within which specific rules apply when accessing the mutable state.
```

File: examples/isolated-functions/isolated_functions.description [hunk 9]
```diff
@@ -1,4 +1,6 @@
...
+// An `isolated` function can access mutable state that is passed as a parameter.
```
Comment (praneesha) [inline examples/isolated-functions/isolated_functions.description:2]:
```suggestion
// An `isolated` function can access a mutable state that is passed as a parameter.
```

File: examples/isolated-functions/isolated_functions.description [hunk 10]
```diff
@@ -1,4 +1,6 @@
...
+// Such a function can access non-`isolated` module-level state only if the variable is `final` and
```
Comment (praneesha) [inline examples/isolated-functions/isolated_functions.description:3]:
```suggestion
// Such a function can access a non-`isolated`, module-level state only if the variable is `final` and
```

File: examples/isolated-functions/isolated_functions.description [hunk 11]
```diff
@@ -1,4 +1,6 @@
...
+// the type is a subtype of `readonly` or `isolated object {}`.
+// An `isolated` function can access `isolated` module-level variables given that they do so within a `lock` statement.
```
Comment (praneesha) [inline examples/isolated-functions/isolated_functions.description:5]:
```suggestion
// An `isolated` function can access `isolated` module-level variables given that they do so within a `lock` statement.
```
"They do so" part is not clear.

File: examples/isolated-objects/isolated_objects.bal [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    decimal longitude;
+|};
+
+isolated class Cities {
+
+    final string country;
+
+    // The initialization expression of a field of an
+    // `isolated` object has to be an `isolated` expression,
+    // which itself will be an isolated root.
```
Comment (praneesha) [inline examples/isolated-objects/isolated_objects.bal:14]:
```suggestion
    // which itself will be an `isolated` root.
```

File: examples/isolated-objects/isolated_objects.bal [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    // e.g., expressions with an immutable static type,
+    // simple literals, constructor expressions
```
Comment (praneesha) [inline examples/isolated-objects/isolated_objects.bal:16]:
```suggestion
    // simple literals and constructor expressions
```

File: examples/isolated-objects/isolated_objects.bal [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+    // where all the subexpressions are isolated expressions, etc.
+    private map<Coordinates> cities = {};
+
+    function init(string country) {
+        self.country = country;
+    }
+
+    function add(string city, Coordinates coordinates) {
+        lock {
+            // Where `self` is being used to access a mutable private
```
Comment (praneesha) [inline examples/isolated-objects/isolated_objects.bal:26]:
```suggestion
            // Where `self` is being used to access a mutable, private
```

File: examples/isolated-objects/isolated_objects.bal [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+            // field, the access needs to be within a `lock` statement.
+            if self.cities.hasKey(city) {
+                io:println("An entry already exists for ", city);
+                return;
+            }
+
+            // Any values transferred into the `lock` statement (and thereby
```
Comment (praneesha) [inline examples/isolated-objects/isolated_objects.bal:33]:
```suggestion
            // Any values transferred into the `lock` statement (and thereby,
```

File: examples/isolated-objects/isolated_objects.bal [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+            // the object) also need to be isolated expressions, ensuring that
```
Comment (praneesha) [inline examples/isolated-objects/isolated_objects.bal:34]:
```suggestion
            // the object) also need to be isolated expressions ensuring that
```

File: examples/isolated-objects/isolated_objects.bal [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+            // the object will continue to be an isolated root.
```
Comment (praneesha) [inline examples/isolated-objects/isolated_objects.bal:35]:
```suggestion
            // the object will continue to be an `isolated` root.
```

File: examples/isolated-objects/isolated_objects.bal [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+            // A `.clone()` or `.cloneReadOnly()` call is an isolated
```
Comment (praneesha) [inline examples/isolated-objects/isolated_objects.bal:36]:
```suggestion
            // A `.clone()` or `.cloneReadOnly()` call is an `isolated`
```

File: examples/isolated-objects/isolated_objects.bal [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+        // `self` can be used outside a `lock` statement if it is being
+        // used to access a `final` field with a type that is a subtype
+        // of `readonly` or `isolated object {}`.
+        io:println("Added coordinates for ", city, ", ", self.country);
+    }
+
+    function getCoordinates(string city) returns Coordinates? {
+        lock {
+            // Similarly, transferring values out of the `lock` statement (and
+            // thereby the object), either using a `return` statement or by
```
Comment (praneesha) [inline examples/isolated-objects/isolated_objects.bal:50]:
```suggestion
            // thereby, the object) either using a `return` statement or by
```

File: examples/isolated-objects/isolated_objects.bal [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+            // assigning to a variable defined outside the `lock` statement,
```
Comment (praneesha) [inline examples/isolated-objects/isolated_objects.bal:51]:
```suggestion
            // assigning to a variable defined outside the `lock` statement
```

File: examples/isolated-objects/isolated_objects.bal [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+            // is also constrained. The expressions transferring values out
```
Comment (praneesha) [inline examples/isolated-objects/isolated_objects.bal:52]:
```suggestion
            // is also constrained. The expressions that are transferring values out
```

File: examples/isolated-objects/isolated_objects.bal [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+            // need to be isolated expressions.
```
Comment (praneesha) [inline examples/isolated-objects/isolated_objects.bal:53]:
```suggestion
            // need to be `isolated` expressions.
```

File: examples/isolated-objects/isolated_objects.bal [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+            return self.cities[city].clone();
+        }
+    }
+
+    function getAllCoordinates() returns string[] {
+        lock {
+            // A method accessing `self` to access a mutable field can
```
Comment (praneesha) [inline examples/isolated-objects/isolated_objects.bal:60]:
```suggestion
            // A method, which accesses the `self` to access a mutable field can
```

File: examples/isolated-objects/isolated_objects.description [hunk 24]
```diff
@@ -0,0 +1,12 @@
+// An `isolated` object's mutable state is isolated from the rest of the program. The object's
+// mutable state is accessible only via the object itself, making it an "isolated root".
```
Comment (praneesha) [inline examples/isolated-objects/isolated_objects.description:2]:
```suggestion
// mutable state is accessible only via the object itself making it an `isolated root`.
```

File: examples/isolated-objects/isolated_objects.description [hunk 25]
```diff
@@ -0,0 +1,12 @@
...
+// Any field in an `isolated` object that is not `final` or is not a subtype of `readonly` or
+// `isolated object {}` has to be a `private` field. Such a field can only be accessed within a
+// `lock` statement, ensuring that there will be no data race when accessing the mutable state of
```
Comment (praneesha) [inline examples/isolated-objects/isolated_objects.description:5]:
```suggestion
// `lock` statement ensuring that there will be no data race when accessing the mutable state of
```

File: examples/isolated-objects/isolated_objects.description [hunk 26]
```diff
@@ -0,0 +1,12 @@
...
+// an `isolated` object.
+// If a `lock` statement in a method accesses such a field, additional rules apply to how values are
+// transferred in and out of the `lock` statement in order to ensure that there are no references
```
Comment (praneesha) [inline examples/isolated-objects/isolated_objects.description:8]:
```suggestion
// transferred in and out of the `lock` statement, in order to ensure that there are no references
```

File: examples/isolated-variables/isolated_variables.bal [hunk 27]
```diff
@@ -0,0 +1,72 @@
+import ballerina/io;
+
+type Coordinates record {|
+    decimal latitude;
+    decimal longitude;
+|};
+
+// The initialization expression of an `isolated` variable
+// has to be an `isolated` expression, which itself will be
+// an isolated root.
```
Comment (praneesha) [inline examples/isolated-variables/isolated_variables.bal:10]:
```suggestion
// an `isolated` root.
```

File: examples/isolated-variables/isolated_variables.bal [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+// e.g., expressions with an immutable static type,
```
Comment (praneesha) [inline examples/isolated-variables/isolated_variables.bal:11]:
```suggestion
// E.g., expressions with an immutable static type,
```

File: examples/isolated-variables/isolated_variables.bal [hunk 29]
```diff
@@ -0,0 +1,72 @@
...
+// simple literals, constructor expressions
```
Comment (praneesha) [inline examples/isolated-variables/isolated_variables.bal:12]:
```suggestion
// simple literals, and constructor expressions
```

File: examples/isolated-variables/isolated_variables.bal [hunk 30]
```diff
@@ -0,0 +1,72 @@
...
+// where all the subexpressions are isolated expressions, etc.
```
Comment (praneesha) [inline examples/isolated-variables/isolated_variables.bal:13]:
```suggestion
// in which all the subexpressions are `isolated` expressions, etc.
```

File: examples/isolated-variables/isolated_variables.bal [hunk 31]
```diff
@@ -0,0 +1,72 @@
...
+    lock {
+        // An `isolated` variable can be accessed within a `lock` statement.
+        if cities.hasKey(city) {
+            io:println("An entry already exists for ", city);
+            return;
+        }
+
+        map<Coordinates> updatedCities = cities.clone();
+        // Any values transferred into the `lock` statement also need to be
+        // isolated expressions, ensuring that the variable will continue
```
Comment (praneesha) [inline examples/isolated-variables/isolated_variables.bal:27]:
```suggestion
        // isolated expressions ensuring that the variable will continue
```

File: examples/isolated-variables/isolated_variables.bal [hunk 32]
```diff
@@ -0,0 +1,72 @@
...
+        // to be an isolated root.
```
Comment (praneesha) [inline examples/isolated-variables/isolated_variables.bal:28]:
```suggestion
        // to be an `isolated` root.
```

File: examples/isolated-variables/isolated_variables.bal [hunk 33]
```diff
@@ -0,0 +1,72 @@
...
+        updatedCities[city] = coordinates.clone();
+        cities = updatedCities.clone();
+    }
+
+    io:println("Added coordinates for ", city);
+}
+
+function getCoordinates(string city) returns Coordinates? {
+    lock {
+        // Similarly, transferring values out of the `lock` statement,
```
Comment (praneesha) [inline examples/isolated-variables/isolated_variables.bal:38]:
```suggestion
        // Similarly, transferring values out of the `lock` statement
```

File: examples/isolated-variables/isolated_variables.bal [hunk 34]
```diff
@@ -0,0 +1,72 @@
...
+        // either using a `return` statement or by assigning to a variable
+        // defined outside the `lock` statement, is also
```
Comment (praneesha) [inline examples/isolated-variables/isolated_variables.bal:40]:
```suggestion
        // defined outside the `lock` statement is also
```

File: examples/isolated-variables/isolated_variables.bal [hunk 35]
```diff
@@ -0,0 +1,72 @@
...
+        // constrained. The expressions transferring values out need to be
```
Comment (praneesha) [inline examples/isolated-variables/isolated_variables.bal:41]:
```suggestion
        // constrained. The expressions, which transfer values out need to be
```

File: examples/isolated-variables/isolated_variables.description [hunk 36]
```diff
@@ -0,0 +1,10 @@
+// An `isolated` variable's mutable state is isolated from the rest of the program. The
+// mutable state reachable from such a variable is only reachable from the variable itself,
```
Comment (praneesha) [inline examples/isolated-variables/isolated_variables.description:2]:
```suggestion
// mutable state is reachable from such a variable only from the variable itself
```

File: examples/isolated-variables/isolated_variables.description [hunk 37]
```diff
@@ -0,0 +1,10 @@
...
+// making an `isolated` variable an "isolated root", similar to `isolated` objects.
```
Comment (praneesha) [inline examples/isolated-variables/isolated_variables.description:3]:
```suggestion
// making an `isolated` variable an "isolated root" similar to `isolated` objects.
```

File: examples/isolated-variables/isolated_variables.description [hunk 38]
```diff
@@ -0,0 +1,10 @@
...
+// Such a variable can only be accessed within a `lock` statement, ensuring that there
```
Comment (praneesha) [inline examples/isolated-variables/isolated_variables.description:4]:
```suggestion
// Such a variable can only be accessed within a `lock` statement ensuring that there
```

File: examples/isolated-variables/isolated_variables.description [hunk 39]
```diff
@@ -0,0 +1,10 @@
...
+// will be no data race when accessing the mutable state reachable from the variable.
+// If a `lock` statement in a function accesses such a field, additional rules apply to how values are
+// transferred in and out of the `lock` statement, similar to `isolated` objects.
```
Comment (praneesha) [inline examples/isolated-variables/isolated_variables.description:7]:
```suggestion
// transferred in and out of the `lock` statement similar to `isolated` objects.
```

File: examples/isolated-variables/isolated_variables.description [hunk 40]
```diff
@@ -0,0 +1,10 @@
...
+// The initial value expression of the variable also needs to be an isolated expression.
```
Comment (praneesha) [inline examples/isolated-variables/isolated_variables.description:8]:
```suggestion
// The initial value expression of the variable also needs to be an `isolated` expression.
```

File: examples/isolated-objects/isolated_objects.bal (hunk 12 above)
Comment (MaryamZi) [reply examples/isolated-objects/isolated_objects.bal:14]: "isolated root" and "isolated expression" are concepts that do not have a direct representation in the code, unlike `isolated` objects and variables.

Applicable to other suggestions for the same too.

File: examples/isolated-objects/isolated_objects.bal (hunk 13 above)
Comment (MaryamZi) [reply examples/isolated-objects/isolated_objects.bal:16]: We have "etc." at the end, is this change correct?

File: examples/isolated-objects/isolated_objects.bal (hunk 14 above)
Comment (MaryamZi) [reply examples/isolated-objects/isolated_objects.bal:26]: It is the private field that is mutable. Do we need commas?

File: examples/isolated-variables/isolated_variables.bal (hunk 29 above)
Comment (MaryamZi) [reply examples/isolated-variables/isolated_variables.bal:12]: Same as previous comment, we're using "etc." at the end.

File: examples/isolated-variables/isolated_variables.description (hunk 36 above)
Comment (MaryamZi) [reply examples/isolated-variables/isolated_variables.description:2]: I'm not sure if this change conveys the original message.

File: examples/isolated-objects/isolated_objects.bal (hunk 23 above)
Comment (MaryamZi) [reply examples/isolated-objects/isolated_objects.bal]: I've updated this a bit. 

```
            // A method which accesses `self`, other than to access a `final`
            // field of a type that is a subtype of `readonly` or
            // `isolated object {}`, can call another function/method only
            // if that function/method is `isolated`.
```

We can't really say the `self` here. `self` is like a variable name here.

Comment (praneesha) [reply examples/isolated-objects/isolated_objects.bal]: We need to say "the" for all nouns even if it is a variable. To be clearer, we can say "the `self` variable".

File: examples/isolated-variables/isolated_variables.description (hunk 36 above)
Comment (praneesha) [reply examples/isolated-variables/isolated_variables.description:2]: Having the "reachable" occurrence doesn't sound correct in this sentence.

File: examples/isolated-variables/isolated_variables.description (hunk 40 above)
Comment (praneesha) [reply examples/isolated-variables/isolated_variables.description:8]: This is not resolved.

File: examples/isolated-functions/isolated_functions.bal (hunk 2 above)
Comment (praneesha) [reply examples/isolated-functions/isolated_functions.bal:35]: +1

File: examples/isolated-functions/isolated_functions.bal (hunk 3 above)
Comment (praneesha) [reply examples/isolated-functions/isolated_functions.bal:36]: The "if it does so" part is not clear. Shall we rephrase the sentence?

Comment (MaryamZi) [reply examples/isolated-functions/isolated_functions.bal]: Rewrote this as

```
// An `isolated` variable that can be accessed within
// a `lock` statement in an `isolated` function.
```

File: examples/isolated-objects/isolated_objects.bal (hunk 17 above)
Comment (MaryamZi) [reply examples/isolated-objects/isolated_objects.bal:35]: See https://github.com/ballerina-platform/ballerina-distribution/pull/954/files#r530761269

File: examples/isolated-objects/isolated_objects.bal (hunk 18 above)
Comment (MaryamZi) [reply examples/isolated-objects/isolated_objects.bal:36]: See https://github.com/ballerina-platform/ballerina-distribution/pull/954/files#r530761269

File: examples/isolated-variables/isolated_variables.bal (hunk 27 above)
Comment (MaryamZi) [reply examples/isolated-variables/isolated_variables.bal:10]: See https://github.com/ballerina-platform/ballerina-distribution/pull/954/files#r530761269

File: examples/isolated-variables/isolated_variables.bal (hunk 32 above)
Comment (MaryamZi) [reply examples/isolated-variables/isolated_variables.bal:28]: See https://github.com/ballerina-platform/ballerina-distribution/pull/954/files#r530761269

File: examples/isolated-variables/isolated_variables.description (hunk 40 above)
Comment (MaryamZi) [reply examples/isolated-variables/isolated_variables.description:8]: See See https://github.com/ballerina-platform/ballerina-distribution/pull/954/files#r530761269

File: examples/isolated-functions/isolated_functions.bal (hunk 3 above)
Comment (praneesha) [reply examples/isolated-functions/isolated_functions.bal]: +1

File: examples/isolated-objects/isolated_objects.bal (hunk 17 above)
Comment (praneesha) [reply examples/isolated-objects/isolated_objects.bal:35]: Ack. +1

File: examples/isolated-variables/isolated_variables.description (hunk 40 above)
Comment (praneesha) [reply examples/isolated-variables/isolated_variables.description:8]: Ack. +1

File: examples/isolated-variables/isolated_variables.description (hunk 36 above)
Comment (MaryamZi) [reply examples/isolated-variables/isolated_variables.description:2]: Any other suggestions? I'm not sure how we can convey this without repeating it.

**Review by hasithaa — APPROVED** [review]
Content looks ok.

File: examples/isolated-variables/isolated_variables.description (hunk 36 above)
Comment (praneesha) [reply examples/isolated-variables/isolated_variables.description:2]: " The mutable state is reachable from such a variable only from itself."

Comment (MaryamZi) [reply examples/isolated-variables/isolated_variables.description:2]: The mutable state here is specifically the mutable state that is reachable from the variable. I'm not sure with this change if it is clear what mutable state we are talking about.

Comment (praneesha) [reply examples/isolated-variables/isolated_variables.description:2]: Let's go ahead with how this is defined in the spec.

File: examples/isolated-objects/isolated_objects.bal (hunk 23 above)
Comment (MaryamZi) [reply examples/isolated-objects/isolated_objects.bal]: +1 to say "the `self` variable"

**Review by praneesha — APPROVED** [review]

**Pull request merged**